pkg/git/local_repository.go

```diff
@@ -326,35 +327,84 @@ type PullResult struct {
 	IsUpdated bool
 }
 
-func (localRepo *LocalRepository) Pull(auth AuthMethod) (PullResult, error) {
+func (localRepo *LocalRepository) Pull(auth AuthMethod) (*PullResult, error) {
 	var gitAuth transport.AuthMethod
 	if auth != nil {
 		gitAuth = auth.toGitAuthMethod()
 	}
+
 	var err error
+	originType := unknownOriginType
+	if !localRepo.DryRun {
+		originType, err = localRepo.originType()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	goGit := originType == httpsOriginType
+
 	log.Debug().
 		Str("repo", localRepo.Path()).
+		Bool("go_git", goGit).
 		Bool("dry_run", localRepo.DryRun).
 		Msg("git: pull")
 	if !localRepo.DryRun {
-		err = localRepo.worktree.Pull(&git.PullOptions{
-			RemoteName: OriginRemote,
-			Auth:       gitAuth,
-		})
-	}
-	if err != nil && !errors.Is(err, git.NoErrAlreadyUpToDate) {
-		return PullResult{}, err
+		if goGit {
+			err = localRepo.worktree.Pull(&git.PullOptions{
+				RemoteName: OriginRemote,
+				Auth:       gitAuth,
+			})
+			if err != nil && !errors.Is(err, git.NoErrAlreadyUpToDate) {
+				return nil, err
+			}
+		} else {
+			if stdout, stderr, err := shell.RunCommand(localRepo.Path(), "git", "pull", OriginRemote); err != nil {
+				return nil, fmt.Errorf("git pull failed:\n\tstdout:[%s]\n\tstderr:[%s]\n%s", stdout, stderr, err.Error())
+			}
+		}
+
 	}
-	return PullResult{IsUpdated: err == nil}, nil
+	return &PullResult{IsUpdated: true}, nil
 }
 
 type PushOp struct {
 	Auth       AuthMethod
 	BranchName string
 }
 
-func (localRepo *LocalRepository) Push(op PushOp) error {
+type originType string
+
+var (
+	sshOriginType     originType = "ssh"
+	httpsOriginType   originType = "https"
+	unknownOriginType originType = "unknown"
+)
 
+func (localRepo *LocalRepository) originType() (originType, error) {
+	remotes, err := localRepo.repo.Remotes()
+	if err != nil {
+		return unknownOriginType, fmt.Errorf("failed to list git remotes: %s", err.Error())
```

Comment:
You can the error value directly if you change `%s` to `%v`. This it will also allow to check for type of the wrapped error. It's minor but I don't see any reason not to do it
Applicable for other places where you are wrapping the error

Reply:
done!